include/RAJA/pattern/forall.hpp

```diff
@@ -589,6 +589,43 @@ RAJA_INLINE camp::resources::EventProxy<Res> CallForallIcount::operator()(T cons
 
 }  // namespace detail
 
+//
+// Experimental support for dynamic policy selection
+//
+
+namespace expt
+{
+
+  enum exec_policy {host_seq, host_parallel, device};
+
+  template<typename POLICY_LIST, typename SEGMENT, typename BODY>
```

Comment:
Based on how things are written, it looks like the ordering of the policies doesn't necessarily have to coincide with the `host`, `host-parallel`, or `device` name. I'm guessing that wouldn't be an easy thing to guard against. Unless, you had another version of this where these policies were specifically defined for them within the function. So, the user would just need to provide like the number of threads to use for the device as a template argument and the dynamic policy. 

Additionally, if I was reading through the `camp::list` implementation correctly it looks like if a user provides fewer than 3 policies that a compile failure will occur due to the `static_assert` in the `at` function for `camp::list`. I could see that as being a potential pain point for people if they have different `policy_list` sprinkled through out their code and have to add `#ifdef #else` like guards in places to protect against from when their code was compiled with different options. The easiest solution for a user would probably to have those defined in a header somewhere, but I just figured I'd bring that up.

Reply:
Having a predefined policy list, `loop_exec, omp_parallel_for_exec, and device' would simplify things. To be honest its probably what most users will want to cycle through, the only template argument could be the number of threads for the gpu. Thoughts guys?

Reply:
I agree with @ajkunen - I think having a list of both policies and "indices" is the right way to implement this, and I think is what @trws was suggesting during the meeting. We could also add some checks based on the Platform types compared to the indices (in the case the index was a Platform).

Reply:
Thanks! I'm blanking out on how to create a parameter pack for the indices using camp, but I pushed up a version that essentially travels down from N-1, to 0 checking for the right policy. I can update it once I find the syntax.